sherpa-onnx/python/sherpa_onnx/utils.py

```diff
@@ -0,0 +1,72 @@
+from typing import Dict, List, Optional
+
+
+def encode_contexts(
+    modeling_unit: str,
+    contexts: List[str],
+    sp: Optional["SentencePieceProcessor"] = None,
+    tokens_table: Optional[Dict[str, int]] = None,
+) -> List[List[int]]:
+    """
+    Encode the given contexts (a list of string) to a list of a list of token ids.
+
+    Args:
+      modeling_unit:
+        The valid values are bpe, char, bpe+char
+      contexts:
+        The given contexts list (a list of string).
+      sp:
+        An instance of SentencePieceProcessor.
+      tokens_table:
+        The tokens_table containing the tokens and the corresponding ids.
+    Returns:
+      Return the contexts_list, it is a list of a list of token ids.
+    """
+    contexts_list = []
+    if "bpe" in modeling_unit:
+        assert sp is not None
+    if "char" in modeling_unit:
+        assert tokens_table is not None
+        assert len(tokens_table) > 0, len(tokens_table)
+
+    if "char" == modeling_unit:
+        for context in contexts:
+            ids = [
+                tokens_table[txt] if txt in tokens_table else tokens_table["<unk>"]
```

Comment:
Shall we consider the case when `txt` is a space?

Reply:
I think for `bpe` and `char` based models, there will not be spaces in context phrases, no?

Reply:
Is "hello world" a valid phrase in this case? 

If not, could you give an example?

Reply:
I think "hello world" will tokenized to something like "hel, lo, _wor, ld" if using bpe, space will be represented as "_" (a unicode version underline).

Reply:
what if the modelling unit is `char`?

Reply:
For a Chinese phrase, there is no spaces at all.

Reply:
I mean for non-CJK languages, e.g., English, German, French, etc.

Reply:
OK, I think there is no need to support all the languages, modeling units for now, the fact is we only have bpe models for English, char models for Chinese and bpe+char models for English+Chinese, for all these models we don't need a space. We could surely add `space` if some models need it, this is just a util function that converts phrases to token ids, I'd rather keep it as simple as it is, we can modify it any time when needed, won't affect anything.

Reply:
Then please document that
- char is for CJK languages, not for English like languages
- BPE is for English-like languages
- bpe+char is for CJK + English-like languages.

Reply:
Ok, I think documenting "char is for CJK language" is enough, for bpe system we don't care the languages.

Reply:
Actually, if you handle spaces, then `char` also works non-CJK languages.

Reply:
Ok,  I will map `space` to `blank`.

Reply:
If you are not willing to make a change, please add
```python3
assert ' ' not in context
```
Otherwise, users will get undesired results without notice
if their modelling unit is `character` and give a phrase
`hello world`.